### PR TITLE
feat: combine heterogeneous data sources

### PR DESCRIPTION
**Pull request by palpatim**
#### Description of changes

Building on the data source refactors in #2092 and #2101, this change enables support for multiple heterogeneous data sources. Supported combinations:

- Any number (up to system limits) of data sources using any strategy: Dynamo or SQL
- Schema models may be related amongst DynamoDB data sources (either default provisioning strategy or Amplify-managed table strategy), and amongst SQL data sources of either MySQL or Postgres. 
- Custom SQL statements

Explicitly not yet supported:
- Relationships across the DynamoDB/SQL boundary
- Many-to-many relationships involving SQL data sources

Known issues:
- `There can be only one type named "Query".` error when declaring custom Query SQL in multiple definitions. Working on a fix for that.

#### Implementation

The primary functional change in this PR is that we know name SQL data sources and associated resources with the name of the strategy, instead of a hardcoded string constant. This allows us to create one stack and associated resources per SQL strategy.

A related change is that we now only create one instance of `RdsModelResourceGenerator` in the model transformer. This one instance will create resources for all SQL data types in one pass, regardless of underlying database type.

#### API

```typescript
AmplifyGraphqlDefinition.combine(definitions: IAmplifyGraphqlDefinition[]): IAmplifyGraphqlDefinition
```

#### Additional changes in this PR

- Opportunistic eslint fixes
- Addressed PR feedback from #2101: strongly typed the "fix up" method, created a utility for easier strategy creation, and created a few test schema snippets to make reuse easier and more predictable

**Note**: SQL Data sources may not be reused across definitions. The correct way to accomplish the use case, “As a customer, I can specify a SQL-backed GraphQL API with schema definitions spread across multiple files”, will be to use the `fromFilesAndStrategy` factory method to create a single definition with multiple files.

#### Description of how you validated changes

- Manual testing pass
- Unit tests pass
- Mock E2E pass
- [Cloud E2E running](https://us-east-1.console.aws.amazon.com/codesuite/codebuild/594813022831/projects/amplify-category-api-e2e-workflow/batch/amplify-category-api-e2e-workflow:e8fcc3e9-c84b-4f0f-8511-98654c3f9aca?region=us-east-1)
- New Cloud E2Es in development, coming in future PR

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by palpatim — COMMENTED** [review]
marker review

File: packages/amplify-graphql-schema-generator/src/__tests__/schema-overrides.test.ts
Comment (sundersc) [reply packages/amplify-graphql-schema-generator/src/__tests__/schema-overrides.test.ts:341]: GraphQL allows array of one length to be represented as single scalar value without brackets. However when we parse it using the library, we need to handle both scalar and list types.

File: packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts
Comment (phani-srikar) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts:108]: can we have models with DDB and SQL strategies within a definition?

File: packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts
Comment (phani-srikar) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts:100]: This is a bit confusing. We can have non-SQL i.e DDB strategy in multiple definitions but not each SQL strategy must be in it's own definition. Wondering what the high level DX is for these cases.

File: packages/amplify-e2e-tests/src/__tests__/sql-generate-unauth.test.ts
Comment (dpilch) [inline packages/amplify-e2e-tests/src/__tests__/sql-generate-unauth.test.ts:51]: For other reviewers, this change was missed as a part of https://github.com/aws-amplify/amplify-category-api/pull/2098

File: packages/amplify-graphql-model-transformer/src/__tests__/model-transformer.test.ts
Comment (dpilch) [inline packages/amplify-graphql-model-transformer/src/__tests__/model-transformer.test.ts:1685]: Should we change this test to expect to not throw rather than removing it?

File: packages/amplify-graphql-relational-transformer/src/resolver/rds-generator.ts
Comment (dpilch) [inline packages/amplify-graphql-relational-transformer/src/resolver/rds-generator.ts:222]:
```suggestion
      throw new Error('The @belongsTo directive is only supported for SQL data sources.');
```

File: packages/amplify-graphql-relational-transformer/src/resolver/rds-generator.ts
Comment (dpilch) [inline packages/amplify-graphql-relational-transformer/src/resolver/rds-generator.ts:172]:
```suggestion
      throw new Error('The @hasOne directive is only supported for SQL data sources.');
```

File: packages/amplify-graphql-api-construct/src/__tests__/__functional__/combine-synthesis.test.ts
Comment (dpilch) [inline packages/amplify-graphql-api-construct/src/__tests__/__functional__/combine-synthesis.test.ts:501]: This is the `There can be only one type named "Query"` error?

File: packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts
Comment (dpilch) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts:105]: nit: 
```suggestion
            `A SQL-based ModelDataSourceStrategy named ('${strategyName}') cannot be shared between definitions. To specify a SQL-based API with schemas across multiple files, use 'fromFilesAndStrategy'`,
```

File: packages/amplify-graphql-transformer-core/src/utils/resource-name.ts
Comment (palpatim) [reply packages/amplify-graphql-transformer-core/src/utils/resource-name.ts:145]: Good callout, will fix in upcoming PR

File: packages/amplify-graphql-model-transformer/src/__tests__/model-transformer.test.ts
Comment (palpatim) [reply packages/amplify-graphql-model-transformer/src/__tests__/model-transformer.test.ts:1685]: The new combine-synthesis tests do that, so this doesn't add any particular value.

File: packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts:108]: Types must be unique across all definitions, regardless of their data source binding.

File: packages/amplify-graphql-api-construct/src/__tests__/__functional__/combine-synthesis.test.ts
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/__tests__/__functional__/combine-synthesis.test.ts:501]: Correct. Am in process of fixing that now

File: packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts:105]: Hmm. The sentence is, "A name cannot be shared", I was trying to give context on which name was duped. I'll change to `The SQL-based ModelDataSourceStrategy ('${strategyName}') was found in multiple definitions, but a strategy name cannot be shared between definitions. To specify...`

File: packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts:100]: It's unfortunately a bit confusing, which you can see in the [tests](https://github.com/aws-amplify/amplify-category-api/pull/2109/files#diff-08f45b7d33ae7b9e339a473334693ec754019ffcbfd773556ec5cec56fbdddb9R259-R265). 

We can explore a cleaner DX, but since that is an additive change, I'd prefer to wait until we get some customer feedback on it; it's awfully easy to put "convenience" methods in place that end up being cumbersome (I'm lookin' at you, `SQLLambdaModelDataSourceStrategyFactory.fromCustomSqlFiles`!)

File: packages/amplify-graphql-transformer-core/src/utils/resource-name.ts
Comment (palpatim) [reply packages/amplify-graphql-transformer-core/src/utils/resource-name.ts]: Fixed in b5eabcffb

File: packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts]: Fixed in f2774ee62

**Review by dpilch — APPROVED** [review]

**Review by svidgen — APPROVED** [review]
Approving API changes, assuming `API.md` diffs are accurate.

Left one comment on a timeout that caught my eye. But, don't mistake this one comment as a signal that I did an in depth review! 😅

File: packages/amplify-e2e-tests/src/__tests__/rds-pg-import.test.ts
Comment (palpatim) [reply packages/amplify-e2e-tests/src/__tests__/rds-pg-import.test.ts:57]: We use this in several places throughout our SQL E2E tests (unfortunately). A retry loop would be my preference, but I didn't want to introduce a new verification paradigm in addition to everything else this series of PRs does. :) I'll add a cleanup task to our backlog

**Pull request merged**
